linkerd/main/src/main/scala/io/buoyant/linkerd/Main.scala

```diff
@@ -34,12 +34,14 @@ object Main extends App {
         val admin = initAdmin(config, linker)
         val telemeters = linker.telemeters.map(_.run())
         val routers = linker.routers.map(initRouter(_))
+        val closableNamers = linker.namers.map(_._2).collect { case x: Closable => x }
 
         log.info("linkerd initialized.")
         registerTerminationSignalHandler(config.admin.flatMap(_.shutdownGraceMs))
         closeOnExit(Closable.sequence(
           Closable.all(routers: _*),
           Closable.all(telemeters: _*),
+          Closable.all(closableNamers: _*),
```

Comment:
On thinking it through I think closable namer should complete before we try and shutdown the router etc.. I did not realize that you had kept closableNamer was synchronous.

I would recommend moving closableNamer to the beginning of the list of closables. This way while we are shutting down the router we can be assured that there are no new requests arriving.